src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,414 @@
+
+  const auto contains_value = lower_bound != INVALID_VALUE_ID && dictionary_values[lower_bound] == search_value;
+
+  // Determine which size attribute vector we have and run the scan on the correct type to avoid virtual calls for
+  // each value access to BaseAttributeVector.
+  const auto small_attribute_vector =
+      std::dynamic_pointer_cast<const FittedAttributeVector<uint8_t>>(dictionary_column.attribute_vector());
+  if (small_attribute_vector) {
+    TableScanImpl<uint8_t> scan;
+    scan.perform_scan(scan_type, chunk_id, small_attribute_vector->values(), static_cast<uint8_t>(lower_bound),
+                      positions, contains_value, pos_list_context);
+    return;
+  }
+
+  const auto medium_attribute_vector =
+      std::dynamic_pointer_cast<const FittedAttributeVector<uint16_t>>(dictionary_column.attribute_vector());
+  if (medium_attribute_vector) {
+    TableScanImpl<uint16_t> scan;
+    scan.perform_scan(scan_type, chunk_id, medium_attribute_vector->values(), static_cast<uint16_t>(lower_bound),
+                      positions, contains_value, pos_list_context);
+    return;
+  }
+
+  const auto large_attribute_vector =
+      std::dynamic_pointer_cast<const FittedAttributeVector<uint32_t>>(dictionary_column.attribute_vector());
+  if (large_attribute_vector) {
+    TableScanImpl<uint32_t> scan;
+    scan.perform_scan(scan_type, chunk_id, large_attribute_vector->values(), lower_bound, positions, contains_value,
+                      pos_list_context);
+    return;
+  }
+
+  throw std::logic_error("Unknown column type provided");
+}
+
+template <typename T>
+void TableScanImpl<T>::perform_scan(const ScanType scan_type, const ChunkID chunk_id,
+                                    const std::vector<T>& attribute_vector, const T& search_value, PosList& positions,
+                                    const bool contains_value, std::optional<PosListContext> pos_list_context) {
+  TableScanContext context = {attribute_vector, search_value, contains_value, chunk_id, pos_list_context};
+
+  // Some special methods are called if there are edge cases or optimisations possible.
+  // Else, default back to comparing with a lambda.
+  switch (scan_type) {
+    case ScanType::OpEquals:
+      return _execute_equals_scan(context, positions);
+    case ScanType::OpNotEquals:
+      return _execute_not_equal_scan(context, positions);
+    case ScanType::OpLessThan: {
+      auto lt = [](const T& val, const T& search_val) { return val < search_val; };
+      return _execute_compare_scan(context, positions, lt);
+    }
+    case ScanType::OpLessThanEquals:
+      return _execute_less_than_equals_scan(context, positions);
+    case ScanType::OpGreaterThan:
+      return _execute_greater_than_scan(context, positions);
+    case ScanType::OpGreaterThanEquals: {
+      auto geq = [](const T& val, const T& search_val) { return val >= search_val; };
+      return _execute_compare_scan(context, positions, geq);
+    }
```

Comment:
Why don't you wrap every case in a method? Some optimizations seem to be applicable for all cases, even if you are still using `_execute_compare_scan` and lambdas.